### PR TITLE
Feature/rr test development

### PR DESCRIPTION
**Pull request by Akustav**

### CODE REVIEW

File: rr_core/include/temoto_resource_registrar/rr_base.h [hunk 1]
```diff
@@ -18,29 +18,39 @@
 #define TEMOTO_RESOURCE_REGISTRAR__RR_BASE_H
 
 #include "rr_client_base.h"
+#include "rr_message_registry.h"
+#include "rr_query_base.h"
 #include "rr_registry.h"
+#include "rr_resource.h"
```
Comment (RValner) [inline rr_core/include/temoto_resource_registrar/rr_base.h]: Seems to be unused and also conceptually it's not going to be used.

File: rr_core/include/temoto_resource_registrar/rr_base.h [hunk 2]
```diff
@@ -18,29 +18,39 @@
 #define TEMOTO_RESOURCE_REGISTRAR__RR_BASE_H
 
 #include "rr_client_base.h"
+#include "rr_message_registry.h"
+#include "rr_query_base.h"
 #include "rr_registry.h"
```
Comment (RValner) [inline rr_core/include/temoto_resource_registrar/rr_base.h]: Are "rr_registry" and "rr_message_registry" functionally different? Also rename "registry" to "catalog"

File: rr_core/include/temoto_resource_registrar/rr_client_base.h [hunk 3]
```diff
@@ -17,23 +17,27 @@
...
-  private:
-    RrRegistryPtr rr_registry_;
+    virtual std::string id();
+
+  protected:
     std::string name_;
+
+  private:
+    std::string id_;
+    //RrRegistryPtr rr_registry_;
```
Comment (RValner) [inline rr_core/include/temoto_resource_registrar/rr_client_base.h]: Conceptually all client and server queries should meet at the RR Catalog, hence there should be a pointer to a RR Catalog

File: rr_core/include/temoto_resource_registrar/rr_identifiable.h [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+ * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */
+
+#ifndef TEMOTO_RESOURCE_REGISTRAR__RR_IDENTIFIABLE_H
+#define TEMOTO_RESOURCE_REGISTRAR__RR_IDENTIFIABLE_H
+
+#include "string"
+
+namespace temoto_resource_registrar
+{
+  class Identifiable
```
Comment (RValner) [inline rr_core/include/temoto_resource_registrar/rr_identifiable.h:25]: What is the purpose of this class?

File: rr_core/include/temoto_resource_registrar/rr_message_registry.h [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+namespace temoto_resource_registrar
+{
+
+  class HashFn
+  {
+  public:
+    std::size_t operator()(const RrQueryRequest &r) const;
+  };
+
+  class RrMessageRegistry
```
Comment (RValner) [inline rr_core/include/temoto_resource_registrar/rr_message_registry.h]: This class should hold both the queries of the clients and servers AND the dependencies between server and client queries

File: rr_core/include/temoto_resource_registrar/rr_registry.h [hunk 6]
```diff
@@ -17,15 +17,84 @@
...
+    std::function<void(RrClientBase)> status_callback_;
+  };
+
+  class RrClientRepository
+  {
+  public:
+    bool add(std::unique_ptr<RrClientBase> client);
+    bool remove(const std::string &id);
+    bool exists(const std::string &id);
+    void registerStatusCallback(std::function<void(RrClientBase)> rrStatusCallback);
```
Comment (RValner) [inline rr_core/include/temoto_resource_registrar/rr_registry.h]: There is no point in having a global status callback registration method. Yes, that's how it used to be but it makes more sense that each "RrClient" object has a status callback which is provided while creating the client

**Review by RValner — CHANGES_REQUESTED** [review]
High level comments:
- Renaming and merging:
  * rename `rr_message_registry` to `rr_catalog`
  * merge `rr_registry` under `rr_base`, because they are supposed to be the same thing
- RrClientRepository and RrServerReporitory seem to share/duplicate code. What I propose is: 
  * define a common templated class called `template<T> class RrServerClientUtil`
  * define `class RrClientRepository : public RrRepository<RrClientRepositoryEntry>`
  * define `class RrServerRepository : public RrRepository<RrServerRepositoryEntry>`
  * move all these util classes under rr_server_client_util.h/.cpp
  so that the common functionalities are defined in `RrServerClientUtil` and the specific functionalities are in the according derived classes.
- There is no point in having a global status callback registration method, as there is in the "RrRegistry" class. Yes, that's how it used to be but it makes more sense that each "RrClient" object has a status callback which is provided while creating the client
- Each client status callback should be part of the client object. Atm the clients and callbacks are in separate maps
- It's currently hard to grasp the motivation for some classes, such as "Identifiable" - so please do add short doxygen styled comments (via doxygen plugin in VS code) in front of the class declarations
- RrClientBase does not have a pointer to the RrCatalog, which seems to imply that the catalog cannot maintain resource dependencies
- QueryRequest and QueryResponse classes should prolly be templated, atm they hold a "std::string", but the requests and responses are unique per server. This is more of a loose subject at the moment, since there is the matter of RR Catalog serialization. TBD.

File: rr_core/include/temoto_resource_registrar/rr_identifiable.h (hunk 4 above)
Comment (Akustav) [reply rr_core/include/temoto_resource_registrar/rr_identifiable.h:25]: This is a interface that describes a method that should be used when storing these in maps. It is their unique identifier

File: rr_core/include/temoto_resource_registrar/rr_base.h (hunk 2 above)
Comment (Akustav) [reply rr_core/include/temoto_resource_registrar/rr_base.h]: Conceptually they are the same, one stores clients and servers, the other stores messages. Both can be merged into one, if need be

File: rr_core/include/temoto_resource_registrar/rr_message_registry.h (hunk 5 above)
Comment (Akustav) [reply rr_core/include/temoto_resource_registrar/rr_message_registry.h]: Dependencies meaning that if a resource is dependent on multiple subresources this relationship is marked down in a map of vectors, for example?

**Pull request closed without merging**